src/amoc_config/amoc_config.hrl

```diff
@@ -0,0 +1,60 @@
+%%==============================================================================
+%% Copyright 2020 Erlang Solutions Ltd.
+%% Licensed under the Apache License, Version 2.0 (see LICENSE file)
+%%==============================================================================
+
+
+%%------------------------------------------------------------------------------
+%% basic types
+%%------------------------------------------------------------------------------
+-type name() :: atom().
+-type value() :: any().
+
+%%------------------------------------------------------------------------------
+%% runtime supplied settings
+%%------------------------------------------------------------------------------
+-type parameter() :: {name(), value()}.
+-type settings() :: [parameter()].
+
+%%------------------------------------------------------------------------------
+%% error types
+%%------------------------------------------------------------------------------
+-type reason() :: any().
+-type error_type() :: atom().
+-type error() :: {error, error_type(), reason()}.
+
+%%------------------------------------------------------------------------------
+%% module attributes definition
+%%------------------------------------------------------------------------------
+-type verification_fun() :: fun((Value :: value()) -> boolean() |
+                                                      {true, NewValue :: value()} |
+                                                      {false, reason()}).
+
+-type update_fun() :: fun((ParamName :: name(), NewValue :: value()) -> any()).
+
+-type maybe_verification_fun() :: verification_fun() | fun((_)-> any()).
+-type maybe_update_fun() :: verification_fun() | fun((_,_)-> any()).
+
+-record(module_parameter, {name :: name(),
+                           mod :: module(),
+                           value :: value(),
+                           verification_fn :: maybe_verification_fun(),
+                           update_fn = read_only :: maybe_update_fun() | read_only}).
+
+-type module_parameter() :: #module_parameter{}.
+-type module_configuration() :: [module_parameter()].
```

Comment:
We call them:
- `required_variables`
- `attribute`
- `parameter`
- `configuration`

Different names suggest it's a different thing, while it is not. I think we need to be more consistent with the naming here, just using `variable`, `attribute` or `parameter` everywhere.

Reply:
`-required_variables(...)` is a module attribute, amoc expects `module_attribute()` type in that attribute.

`parmeter()` is a key/value (`name()` and `value()` types)  tuple, `settings()` is the list of parameters. settings is something that can be provided in `amoc:do/3`/`amoc_dist:do/3` interface or via REST API.

The only types exported from `amoc_config_*` modules are : `name()`, `value()`, `settings()`. if you find `parameter()` type confusing, i can just remove it. it's not used directly anywhere.

`module_parameter()` is internal representation of parsed module attributes, `module_configuration()` is the list of `module_parameter()` elements. here i intentionally used term configuration (not setting) to avoid any confusion in code.

Naming of the variables is done in the same manner. if you see any variable called Param - that one is `module_parameter()`.

there is only one naming inconsistency, `amoc_config_validation` module name while in all other places `verification` term is used. i just didn't want rename the file but i can do it if you insist.